### ISSUE TITLE
Django Build

### ISSUE THREAD

**Issue by asewnath** [JCSDA-internal/evagram#3]
## Description
This part of the evagram development is to take a first pass at the frontend and backend design. Emphasis is on producing bare minimum code to complete this phase. Further full-stack development will be in the next phase.

Process for completing this phase:

- [x] Build Django application with preliminary frontend
- [x] Render plots using React framework
- [x] Link Database (on personal computer) to Django application
- [x] Create unit testing for Django application
- [x] Build Github actions to run unit tests
- [x] Create PR and merge backend/prelim frontend, unit tests, and GitHub actions to eva-gram

## Acceptance Criteria (Definition of Done)

First pass of eva-gram website is complete. Frontend changes on-the-fly as database is updated.

**Comment by brandonzhu09** [comment]
![Evagram-ERD](https://github.com/JCSDA-internal/evagram/assets/38299259/339b5ff9-f0b7-437f-9350-ee67d01c85cc)

**Comment by brandonzhu09** [comment]
![Evagram](https://github.com/JCSDA-internal/evagram/assets/38299259/7af97f54-2c83-4ae5-8f6e-7463addd59e9)

**Comment by asewnath** [comment]
Just as a reminder, the proposed schema modification was to remove the relationship from plots and groups and add a new relationship to groups and variables

**Comment by brandonzhu09** [comment]
Should I go ahead and create the new PR to update the schema with the proposed changes?

**Comment by asewnath** [comment]
@tariqhamzeyssai Have you gotten a chance to think over the proposed changes?

**Comment by ghost** [comment]
I need to see precisely how you are implementing this change.  I'm not sure if I love it enough, or see enough benefit, to justify it.  Did we say the drop-down should cascade like:  `Experiment -> Observation -> Variable -> Group`?  How exactly does this change help?  If you relate `groups` to `variables`, why wouldn't it just be incorporated into the `observation_variables` registry?  Conceptually speaking, it doesn't make much sense to me. 

On that note, I'm having doubts about the `observation_variable` itself.  It stores redundant information that's already encoded in `plots`, which may be violating some database normalization standards.

Please correct me where I'm wrong, thanks

**Comment by asewnath** [comment]
@brandonzhu09 I think we need to have more of a discussion on this database schema, so while we work this out, please continue trying to solve the issue with rendering plots in React. 

@tariqhamzeyssai Yes, the dropdown cascade would be as you said, `Experiment -> Observation -> Variable -> Group`. Let's say we are looking at the observation `aircraft`. In `aircraft`, there are three variables to choose from: `airTemperature`, `windNorthward`, and `windEastward`. Each of these variables will have plots associated with it, which we are calling groups (admittedly not the most descriptive word to choose for it, but it follows the eva paradigm of describing a field according to `experiment::group::variable`. You can think of it as what field is actually visualized). So for `windEastward` it would have groups such as `effectiveerror-vs-gsifinalerror` and `hofx-vs-gsihofx`. These plots represent the typical diagnostics that eva would generate during a swell run. If a user wanted to see all plots available for aircraft/windEastward, they would use drop downs to search for `aircraft -> windEastward` and the final drop down would show all the plots.

In terms of duplicate fields in the database, there's a lot of replication of variables and plots. For example, observation `satwind` also has variables called `windEastward` and `windNorthward` and each variable would have many if not all the same groups (plots) that are found with `aircraft` variables. So the question is how do we organize the schema for repeating variables and groups. This is of course on top of the duplication of the same `observations` found in different `experiments`.

In regards to your thoughts regarding `observation_variable`, I completely agree that it doesn't make much sense. I didn't notice the redundant information, and the schema doesn't look conducive to the drop down flow we ultimately want. Definitely need to think about this more. I think the relationships are a little confused

**Comment by asewnath** [comment]
@tariqhamzeyssai @brandonzhu09 

Keeping the drop down flow in mind, would it make sense for every table to have a one-to-many relationship?

`Owner` -> one-to-many -> `Experiment` -> one-to-many -> `Observation` -> one-to-many -> `Variable` -> one-to-many -> `Group` -> one-to-many -> `Plots` which contains the div and script components. Arrows here indicate the drop down progression

**Comment by brandonzhu09** [comment]
I want to quickly mention that I initially liked the idea of correlating all the tables to the central `plots` table to reduce replication of the other tables and reflected that in the final UML design. However, I found it difficult correlating tables like `groups` and `variables` in the schema as the dropdown cascade suggests there should be a correlation there.

**Comment by asewnath** [comment]
Yeah, I'm wondering if we should just go back to the original schema design and forgo the dropdown cascade for now. The dropdown after experiment would just show all possible observation/variable/group selections. Then, we can continue focusing on getting the django app into evagram. What do you think @tariqhamzeyssai?

**Comment by brandonzhu09** [comment]
I do want to point out that the point of the dropdown cascade is to provide the user with valid input options to choose from and by making all the observation/variable/group selections possible, it will lead to some selections to be an invalid query, and nothing gets displayed. The dropdown cascade was my idea to connect the relationships together, but I am open to other alternatives.

**Comment by ghost** [comment]
Why can't we implement cascading options using the central `plots + 4 appendages` model?  It can be done with a Python function.  All it is is a layered subset query.  Assign each attribute (experiment, group...) a priority number.  Work up/down the priority levels to get the final subset or selection.

**Comment by brandonzhu09** [comment]
@tariqhamzeyssai I was not aware of this practice. I will try to look into it further.

**Comment by asewnath** [comment]
@tariqhamzeyssai That sounds good! Is there an example of this implementation somewhere

**Comment by ghost** [comment]
Hmm... after further consideration this may not be so straightforward to implement. @brandonzhu09  You can try to spend time making sense of this comment, but focus on more immediate concerns first.

In a pure python sense, I'm just saying we can construct our own queries to facilitate whatever cascading behavior we want.  Here's some example code.  Whenever the user selects a drop-down option, this function runs and updates the drop-down values.
```
def subset_plots(**kwargs):

    '''
    The input kwargs contain the name of the drop-down that was altered (selected) as well as the values in all the other drop-downs.
    kwargs = {'selected': attribute_table, 'experiments'=None, 'observations'=None, 'variables'=None, 'groups'=None}
    
    subset order:
    1. experiments
    2. observations
    3. variables
    4. groups
    '''

    return_dict = copy.copy(kwargs)

    # We will construct a multi-part query
    join_clauses = []
    where_clauses = []
    values = []

    priority_list = ['experiments', 'observations', 'variables', 'groups']

    # Loop over the drop-downs until the altered drop-down is reached 
    for ind in range(len(priority_list)):

        table = priority_list[ind]

        join_clauses.append(f"join {table} on...")

        if kwargs[table] is None:
            break

        where_clauses.append(f"where {table}_name=%s")
        values.append(kwargs[table])

        if table == kwargs['selected'] or ind==len(priority_list)-2:
            table = priority_list[ind+1]
            join_clauses.append(f"join {next_table on...")
            break

    query_root = f"select {table}_name from plots"
    full_query = f"{query_root} {' '.join(join_clauses)} {' AND '.join(where_clauses)}"

    results = conn.execute(full_query, tuple(values))

    option_list = unique(results)

    return_dict[table] = option_list

    return return_dict

```
This is not perfect, but does it convey the point?  I'm realizing two things, 1) This is likely very inefficient, and 2) It would need to make an asynchronous request via Ajax.  Alternatively I've facilitated this kind of behavior using javascript listeners operating on a nested dictionary in memory.  Is there a better way to do this in Django that I'm not aware of yet?

**Comment by asewnath** [comment]
@brandonzhu09 for now is it possible to just offer a list of all valid plots after a user selects owner and experiment (and perhaps go back to the original schema?). There's definitely a lot of rabbit holes we can go down in regards to this project, but we have to consider time. I know plot rendering with React is still an issue, which I would classify as more of an immediate concern

Oh, just saw that you checked off the React/render plot box! Great work!! Let's keep moving down the checklist for now

**Comment by ghost** [comment]
@brandonzhu09 @asewnath Which schema is the "original"?

**Comment by asewnath** [comment]
> ![Evagram-ERD](https://private-user-images.githubusercontent.com/38299259/301877497-339b5ff9-f0b7-437f-9350-ee67d01c85cc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcyNDQxMjgsIm5iZiI6MTcwNzI0MzgyOCwicGF0aCI6Ii8zODI5OTI1OS8zMDE4Nzc0OTctMzM5YjVmZjktZjBiNy00MzdmLTkzNTAtZWU2N2QwMWM4NWNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA2VDE4MjM0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM4YzBlZDUzNTg4Nzk4ZWVmMjE0ZGI0NmJkYmIzODA0YjI0MmMxZWI2OWQxMTI1MjlhOTM1NDMxMTllY2NhOGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6VuqaDhFW5ie5ho5rVVRrtEJMUyg-FHijc8eEfc4mic)

This one is the original one I believe

**Comment by brandonzhu09** [comment]
> I need to see precisely how you are implementing this change. I'm not sure if I love it enough, or see enough benefit, to justify it. Did we say the drop-down should cascade like: `Experiment -> Observation -> Variable -> Group`? How exactly does this change help? If you relate `groups` to `variables`, why wouldn't it just be incorporated into the `observation_variables` registry? Conceptually speaking, it doesn't make much sense to me.
> 
> On that note, I'm having doubts about the `observation_variable` itself. It stores redundant information that's already encoded in `plots`, which may be violating some database normalization standards.
> 
> Please correct me where I'm wrong, thanks

I will try my best to clear up any confusion with how the current implementation stands now. The central `plots` table contains the foreign keys of the Experiment, Observation, Variable and Group. As far as querying the database for the plot is concerned, these FKs are all that is required to select the right plot to be displayed to the user. In order for the user to generate a plot, they would need to interact with the drop-down menus with the valid options created from the dropdown cascade. That is where the junction tables come in. The `observation_variable` table is what Django interacts with to get all the valid variables for a single observation (I would refer to Django documentation for the ManyToManyField class for more details). With that said, I can't speak to junction tables and to its redundancy in the `plots` table but with what I have now, Django is properly handling the dropdown cascade with the current implementation, except with `groups` of course.

**Comment by ghost** [comment]
@asewnath I don't see an image with that link.

**Comment by brandonzhu09** [comment]
> @brandonzhu09 for now is it possible to just offer a list of all valid plots after a user selects owner and experiment (and perhaps go back to the original schema?). There's definitely a lot of rabbit holes we can go down in regards to this project, but we have to consider time. I know plot rendering with React is still an issue, which I would classify as more of an immediate concern
> 
> Oh, just saw that you checked off the React/render plot box! Great work!! Let's keep moving down the checklist for now

Thank you! We can probably discuss if there is anything further to be done with rendering the plots later.

**Comment by asewnath** [comment]
> > I need to see precisely how you are implementing this change. I'm not sure if I love it enough, or see enough benefit, to justify it. Did we say the drop-down should cascade like: `Experiment -> Observation -> Variable -> Group`? How exactly does this change help? If you relate `groups` to `variables`, why wouldn't it just be incorporated into the `observation_variables` registry? Conceptually speaking, it doesn't make much sense to me.
> > On that note, I'm having doubts about the `observation_variable` itself. It stores redundant information that's already encoded in `plots`, which may be violating some database normalization standards.
> > Please correct me where I'm wrong, thanks
> 
> I will try my best to clear up any confusion with how the current implementation stands now. The central `plots` table contains the foreign keys of the Experiment, Observation, Variable and Group. As far as querying the database for the plot is concerned, these FKs are all that is required to select the right plot to be displayed to the user. In order for the user to generate a plot, they would need to interact with the drop-down menus with the valid options created from the dropdown cascade. That is where the junction tables come in. The `observation_variable` table is what Django interacts with to get all the valid variables for a single observation (I would refer to Django documentation for the ManyToManyField class for more details). With that said, I can't speak to junction tables and to its redundancy in the `plots` table but with what I have now, Django is properly handling the dropdown cascade with the current implementation, except with `groups` of course.

Thanks for the explanation @brandonzhu09. So the current implementation of the database is the second schema you've sent in this issue? With the observation_variable table?

**Comment by asewnath** [comment]
> @asewnath I don't see an image with that link.

@tariqhamzeyssai Sorry about that, I was pointing to the first schema brandon put in this issue, before the addition of the observation_variable table

**Comment by brandonzhu09** [comment]
@asewnath Yes, that's right

**Comment by asewnath** [comment]
@brandonzhu09 Okay, the issue @tariqhamzeyssai was raising regarding the addition of the observation_variable table was that it was introducing redundancy. The redundancy is that there are observation and variable foreign keys in both the observation_variable and plots tables. Correct me if I'm wrong Tariq.

It seems to me that the question is if that redundancy can be removed while still maintaining the relationship between observations and variables.

**Comment by ghost** [comment]
@brandonzhu09  Can you push the branch for your Django implementation so we can see it?  Maybe some confusion (mostly my confusion!) will be cleared up.  I normally think in raw SQL terms, so I need to see your Django code to knowledgeably discuss it.

Your reasoning for the junction table makes sense, but I will reiterate that `plots` is, effectively, already a 4-way junction table with `observation_variable` encoded.  No?  I will insist that redundancy is to be eliminated.  It's still not quite clear to me why the `plots + 4 appendages` model (which I like for its simplicity and consistency) is insufficient for the cascading drop-down, but Again I need to see your code for this so I can understand better.  Then I will get back to you if I still can't grasp the problem.

As Akira said, just focus on the checklist for this issue, get the app up and running and testing.  We can later revisit the schema as a final refinement before the project wraps up.  Thanks for tolerating my feedback!

**Comment by asewnath** [comment]
@tariqhamzeyssai your feedback is always appreciated!

**Comment by brandonzhu09** [comment]
@tariqhamzeyssai I took a longer look into the `plots + 4 appendages` model and found that I was able to correlate all the tables using `plots` as the junction table! I was able to implement the idea relatively simple with Django field lookups but with the new implementation, I was able to remove the redundant `observation_variable` junction table. Here is the updated schema:

![Evagram-Final](https://github.com/JCSDA-internal/evagram/assets/38299259/fb8ce16e-6cdc-48c5-833c-a0e648aeea86)

**Comment by brandonzhu09** [comment]
What do you guys think is an appropriate name for each of the branches for both the input tool and the Django build?

**Comment by asewnath** [comment]
> What do you guys think is an appropriate name for each of the branches for both the input tool and the Django build?

As long as each branch starts with "feature/", it doesn't matter too much. Maybe "feature/update_schema" for input tool related stuff and "feature/initial_django" for django stuff

**Comment by asewnath** [comment]
@brandonzhu09 To do list (mainly for me to keep track of things)

- [x] PR for modification of input_tool and any other modifications needed for the schema change
- [x] PR for initial Django work
- [x] Draft slides for presentation later this month
